msyslog-v1.08a+smac/src/man/om_udp.8
.\"	$CoreSDI: om_udp.8,v 1.1.2.3 2001/11/20 09:56:23 alejo Exp $
.\"
.\" Copyright (c) 2001
.\"	Core-SDI SA. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of Core-SDI SA nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd Feb 28, 2001
.Dt OM_UDP 8
.Os Core-SDI
.Sh NAME
.Nm msyslog udp
output module
.Nd
.Xr syslogd 8
output module used to forward messages to remote hosts through UDP
.Sh SYNOPSIS
.Nm udp
.Op Fl a
.Op Fl h Ar host
.Op Fl p Ar port
.Sh DESCRIPTION
.Nm udp
output module forwards messages to remote hosts trhoug a UDP connection.
.Sh EXAMPLES
.Bd -literal
# Send all logs to host loghost.somedomainexample.com port syslog (514)
*.* %udp -h loghost.somedomainexample.com -p syslog
# Same, but add host to message string (with -a)
*.* %udp -a -h loghost.somedomainexample.com -p syslog
.Ed
.Sh BUGS
.Bl -bullet
.It
Messages sent through UDP get lost, consider using TCP
.It
We are sending everything not encrypted! You may want to use a tunnel
such as SSL.
.It
Submit bugs at this project's Sourceforge Bug reporting system at:    
http://sourceforge.net/tracker/?func=add&group_id=25741&atid=385117
You may also report them directly to the authors; send an email to
core.devel.alat@corest.com, describing the problem the most you can,
containing also machine description, hardware description, the
configuration file (/etc/syslog.conf), the OS description, and the
invoking command line.
The more you describe the bug, the faster we can fix it.
.El
.Sh SEE ALSO
.Xr syslog 3 ,
.Xr syslog.conf 5 ,
.Xr om_classic 8 ,
.Xr om_mysql 8 ,
.Xr om_peo 8 ,
.Xr om_pgsql 8 ,
.Xr om_regex 8 ,
.Xr om_tcp 8 ,
.Xr peochk 8 ,
.Xr syslogd 8


html/ChangeLog.txt
$Id$

Feature Request:

Bugs:
1.  Change code to allow for tailing spaces when looking at the expression field

3/10/2004
	-  Removing any reference to company name

3/9/2004
	-  Changed out company logo
	-  Added GPL notice to all libraries

1/21/2004
	-  Found a bug where a host would email off an alert with no text.  I've put a stop gap fix in.

1/13/2004
        -  Fixed bug with msyslog.pgsql where SET STORAGE didn't have a space before it.
        -  Fixed bug where hour in time of rules would default to 18:00 for no good reason
	-  Fixed two problems with cloning rules:  1)  order was not preserved  2)  an imported version of the database couldn't clone because fields that were empty needed to have non-null defaults applied
	-  Forgot to merge in latest VPN reports

1/6/2004
        -  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/6/2004
	-  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/1/2004
	-  Fixed spelling error in with the word 'threshold' in the rules schema
	-  Fixed a bug in vacuumdb where it was 'ANALYZ' not 'ANALYZE'

12/19/2003
	-  Updated code to look into /opt/apache

12/10/2003
	-  Updated database to schema to not use compressed text fields(we'll see how this performs)
	-  working on adding interfaces to more lock data and other new stats with PostgreSQL V7.4
	-  cleaned up some button descriptions on the maintenance page, also added a lock view as well as a settings view

12/9/2003
	-  Finished rule.php support for basic timer maintenance.  Need to add another page to graft timers onto rules.
	-  processlogs.php is now setup to support rule timers
	-  need to write rule expiration process 
	-  started converting away from compressed text in hopes of providing faster data retrieval.... also pulled OIDs from DB definition

12/8/2003
	-  Adding support for date and time based rules with date ranges, day of week selections, as well as deleted rules, need to add interface to control rule timestamp properties

12/7/2003
	-  finished adding accumulation thresholds

12/3/2003
	-  adding support for both types of thresholds.  Need to update web pages to reflect new radio buttons.
	-  added web configuration support for supression thresholds and accumulating thresholds, now onto updating the log processor
	-  basic supression works(kinda), need to verify functionality

12/2/2003
	-  customer profiles can now have multiple hosts added at once.
	-  fixed a stupid bug where <prev|refresh|next> didn't behave right.  next would stop working if you started at oct-26-2003 and it would stay on oct-26-2003
	-  Added web-based framework and database schema to support alert supression thresholds

12/1/2003
	-  Updated processlogs.php code to better deal with single entry/no rules vs mutliple rules

11/19/2003
	-  Updated mail table to enforce unique login ids(effectively stopping two processes from running at the same time, one will crash and die(safely))
	-  Updated openmail and closeopenmail to use transaction support since PostgreSQL no longer does server-side auto-commit(ie. convert everything to transactions)
	-  Transaction support should now be officially added, will do some testing

11/18/2003
	-  Updating program for support with PostGreSQL V7.4
	-  Fixed host process table to reflect the fact that multiple hosts are in there by default	

10/13/2003
	-  Added alert total to bottom of alert aggregation

10/6/2003
	-  finished support for alert aggregation
	-  updated 1stview to pull the current time and date
	-  fixed host.php bug where you could expire syslogs but not be forced to expire syslogs

10/5/2003
	-  still adding support for alert aggregation.  Basic aggregation works plus alert zooming but need to add support for across the board for other alert queries

10/3/2003
	-  added ability to un/suspend log processors from the web interface
	-  updated maintenance to rebuild all indexes in an better manner(ie. grab the index list from the DB rather then by hand)
	-  added additional framework to do alert aggregation interface.  Need to add 'aggregation code' for display

9/24/2003
	-  weeklyindexrebuild.php now pulls all indexes from the system and rebuilds them.  The result is that the system will now rebuild any new indexes without manual reconfiguration.

9/15/2003
	-  launchid was not initiated correctly in the clonerule.

9/2/2003
	-  Fixed a bug where '\'s at the end of a line caused problems because we were not properly dealing with them in general.  Fixed that.

8/23/2003	
	-  Created another bug when fixing 5000 line paging.  Timestamp was thrown off in view.php

8/18/2003	
	-  Syncing changes from production smt environment:  vacuumtsyslog.php
	-  Updated weeklyindexrebuild.php to account for the correct indexes
	-  Updated maint.php to account for the three new indexes for the launch program section

8/13/2003
	-  view.php has had several updates.  Paging should now be fixed.  Multiple searches appeared not to be working correctly.
	-  Needed to add lastid as hidden var if the variable was set
	-  Needed to use urlencode on top of htmlspecialcharacters, filters were broken because of it

8/11/2003
	-  1stfilter.php doesn't list 'global' filters that you down own
	-  modified view.php to not let the user save a filter with no description
	-  added support to delete all of a user's filters(ie. do before delete)
	-  another problem popped up with filter.php when I added the delete user filter option

8/1/2003	
	-  processlogs.php now supports launching external programs!

7/31/2003
	-  Updated vacuumdb.php to do a full vacuum of the TSyslog table.  Why?  Because the system doesn't reclaim disk space or use old delete space for some reason
	-  Almost finished adding launch program code, need to test.

7/27/2003
	-  Added weeklyindexrebuild.php which rebuilds all indexes at 5am Sunday morning

7/23/2003

msyslog-v1.08a+smac/src/modules/om_udp.c
/*	$CoreSDI: om_udp.c,v 1.1.2.2 2001/11/30 23:13:45 alejo Exp $	*/
/*
     Copyright (c) 2001, Core SDI S.A., Argentina
     All rights reserved
   
     Redistribution and use in source and binary forms, with or without
     modification, are permitted provided that the following conditions
     are met:
   
     1. Redistributions of source code must retain the above copyright
        notice, this list of conditions and the following disclaimer.
     2. Redistributions in binary form must reproduce the above copyright
        notice, this list of conditions and the following disclaimer in the
        documentation and/or other materials provided with the distribution.
     3. Neither name of the Core SDI S.A. nor the names of its contributors
        may be used to endorse or promote products derived from this software
        without specific prior written permission.
   
     THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
     IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
     OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
     IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
     INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
     NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
     DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
     THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
     (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
     THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 *  om_udp -- UDP output module
 *
 * Author: Alejo Sanchez for Core-SDI SA
 *
 */

#include "config.h"

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/uio.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <syslog.h>

#if TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif

#include "../modules.h"
#include "../syslogd.h"

struct om_udp_ctx {
	int	fd;
	char	*host;
	char	*port;
	void	*addr;	/* remote address */
	int	addrlen;
	int	flags;
};

#define M_ADDHOST	0x01

int sock_udp(char *, char *, void **, int *);
int udp_send(int, char *, int, void *, int);

/*
 *  INIT -- Initialize om_udp
 *
 *  we get remote host and port
 *
 *  usage: -h host (required)
 *         -p port (required)
 *         -a	   (add hostname to message)
 *
 * we try to make it the most IPv6 compatible as we can
 * for future porting
 *
 */

int
om_udp_init(int argc, char **argv, struct filed *f, char *prog, void **ctx,
    char **status)
{
	struct	om_udp_ctx	*c;
	int			ch;

	dprintf(MSYSLOG_INFORMATIVE, "om_udp init: Entering\n");

	if ((*ctx = (void *) calloc(1, sizeof(struct om_udp_ctx))) == NULL) {
		dprintf(MSYSLOG_INFORMATIVE, "om_udp_init: couldn't allocate"
		    " context\n");
		return (-1);
	}
	c = (struct om_udp_ctx *) *ctx;

	/* parse line */
	optind = 1;
#ifdef HAVE_OPTRESET
	optreset = 1;
#endif
	while ((ch = getopt(argc, argv, "h:p:m:s:a")) != -1) {
		switch (ch) {
		case 'h':
			/* get remote host name/addr */
			c->host = strdup(optarg);
			break;
		case 'p':
			/* get remote host port */
			c->port = strdup(optarg);
			break;
		case 'a':
			c->flags |= M_ADDHOST;
			break;
		default:
			dprintf(MSYSLOG_SERIOUS, "om_udp_init: parsing error"
			    " [%c]\n", ch);
			if (c->host)
				free(c->host);
			if (c->port)
				free(c->port);
			free(*ctx);
			return (-1);
		}
	}

	if ( c->host == NULL) {
		dprintf(MSYSLOG_SERIOUS, "om_udp_init: host unspecified\n");
		return (-1);
	}

	errno = 0;
	if ((c->fd = sock_udp(c->host, c->port == NULL? "syslog" : c->port,
	    &c->addr, &c->addrlen)) == -1) {
		dprintf(MSYSLOG_SERIOUS, "om_udp_init: error creating generic"
		    " outgoing UDP socket [%s]", strerror(errno));
		return (-1);
	}

	return (1);
}


/*
 *  WRITE -- Send a message
 *
 */

int
om_udp_write(struct filed *f, int flags, struct m_msg *m, void *ctx)
{
	struct om_udp_ctx *c;
	char time_buf[16];
	char line[MAXLINE + 1];
	int l;

	if (m == NULL || m->msg == NULL || !strcmp(m->msg, "")) {
		logerror("om_udp_write: no message!");
		return (-1);
	}

	c = (struct om_udp_ctx *) ctx;

	strftime(time_buf, sizeof(time_buf), "%b %e %H:%M:%S", &f->f_tm);

	/* we give a newline termination to difference lines, unlike UDP */
	if (c->flags & M_ADDHOST) {
		l = snprintf(line, sizeof(line), "<%d>%.15s %s %s\n",
		    f->f_prevpri, time_buf, f->f_prevhost, m->msg);
	} else {
		l = snprintf(line, sizeof(line), "<%d>%.15s %s\n",
		    f->f_prevpri, time_buf, m->msg);
	}

	dprintf(MSYSLOG_INFORMATIVE, "om_udp_write: sending to %s:%s, %s",
	    c->host, c->port, line);

	if (udp_send(c->fd, line, l, c->addr, c->addrlen) == -1) {

		dprintf(MSYSLOG_SERIOUS, "om_udp_write: error sending "
		    "to remote host [%s]", strerror(errno));
	}
			
	return (1);
}


/*
 *  CLOSE -- close om_udp
 *
 */

int
om_udp_close(struct filed *f, void *ctx)
{
	struct om_udp_ctx *c;

	c = (struct om_udp_ctx *) ctx;

	if (c->host)
		free(c->host);
	if (c->port)
		free(c->port);
	if (c->addr)
		free(c->addr);

	close(c->fd);

	return (1);
}
